File: wordbreak_2.cpp
#include<iostream>
#include<map>
#include<vector>

using namespace std;

vector<string> util(string s, map<string,int> &m, map<string, vector<string> > &d) {
    if (d.count(s) > 0)
        return {s};

    if (s.length() == 0)
        return {};

    string res = "";
    for (const pair <string,int> &ele : m) {
        string eles = ele.first;
        if (eles.length() > s.length())
            continue;

        if (eles != s.substr(0,eles.length()))
            continue;

        if (eles.length() == s.length()) {
            res += eles;
            break;
        }
        else {
            vector<string> rest = util(s.substr(eles.length()), m, d);
            for (const string &trest : rest) {
                res = res + eles + " " + trest;
            }
        }
    }

    d[s].push_back(res);
    return {res};
    
}

void wordBreak(string s, vector<string> &wordDict) {
    map<string, int> m;
    for (const string &w : wordDict) {
        m[w] = 1;
    }
    map<string, vector<string> > d;
    vector<string> res = util(s, m, d);
    for (const string &out : res) {
        cout << out << endl;
    }
}

int main () {
    string s1 = "catsanddog";
    vector<string> w = {"cat", "cats", "and", "sand", "dog"};
    wordBreak(s1,w);
}

File: wordbreak_2.py
def util( s, wordDict, d):
    if s in d:
        return d[s]
    if s == '':
        return []
    res = []
    for w in wordDict:
        if len(s) < len(w):
            continue
        if s[:len(w)] != w:
            continue
        if len(s) == len(w):
            res.append(w)
        else:
            rest = util(s[len(w):], wordDict, d)
            for pat in rest:
                pat = w + " " + pat
                res.append(pat)
    d[s] = res
    return res

def wordBreak(s, wordDict):
    """
    :type s: str
    :type wordDict: List[str]
    :rtype: List[str]
    """
    d = {}
    return util(s, wordDict, d)

s = "catsanddog"
w = ["cat", "cats", "and", "sand", "dog"]
print wordBreak(s,w)
